desugarer.go

```diff
@@ -236,6 +236,11 @@ func desugarArrayComp(comp *ast.ArrayComp, objLevel int) (ast.Node, error) {
 
 func desugarObjectComp(comp *ast.ObjectComp, objLevel int) (ast.Node, error) {
 
+	if objLevel == 0 {
+		dollar := ast.Identifier("$")
+		comp.Fields = append(comp.Fields, ast.ObjectFieldLocalNoMethod(&dollar, &ast.Self{}))
```

Comment:
Does this modify the input AST?  Is that OK?

Reply:
Yes, it does. It's not a problem, because desugaring in both C++ and Go modifies the original AST instead of creating a new one. In particular we append this special field to regular objects, too.

I think immutable AST could be cleaner, especially for public API, but that's a separate discussion (and out of scope for this change).